### PR TITLE
[RFC] Add Sysctl interface support for OSX/Linux/FreeBSD

### PR DESCRIPTION
**Pull request by elthariel**
Hi,

This is an attempt at defining an interface to the `sysctl` API of the various UNIXes. OpenBSD is not yet supported as they don't have an easy way to get a sysctl by name (that I'm aware of). We could implement it by spawning a `sysctl` process or define a static map somewhere.

The underlying goal of this is to have a way to query the number of CPUs on OSX to provide a better default for the number of default threads on the multithread support branch(es). In addition to that we could use it for `System.hostname`, and I think this would be useful for many folks out there.

This has only been tested on OSX yet. I'll soon test it on Linux but I'll need help for FreeBSD. Any comment on this would be greatly appreciated, especially regarding general usefulness, design and method naming.

Unicornly yours,
Lta.

Signed-off-by: Julien 'Lta' BALLET <contact@lta.io>

### CODE REVIEW

File: src/sysctl.cr [hunk 1]
```diff
@@ -0,0 +1,87 @@
+lib LibC
```
Comment (Sija) [inline src/sysctl.cr]: This `lib` declaration IMO should be extracted to a separate file for each architecture (`src/lib_c/<arch>/c/sysctl.cr`).

File: src/sysctl.cr [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      if res == 0
+        String.new(value, value_len.value - 1, 1)
+      else
+        raise Errno.new("Unable to fetch sysctl value #{name}")
+      end
+    {% elsif flag?(:linux) %}
+      path = name_to_path name
+      read_proc_sys(path)
+    {% else %}
+      raise NotImplemented("Sysctl", "is not supported on this platform")
```
Comment (Sija) [inline src/sysctl.cr]: Missing `.new`.

File: src/sysctl.cr [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      if res == 0
+        value.value
+      else
+        raise Errno.new("Unable to fetch sysctl value #{name}")
+      end
+    {% elsif flag?(:linux) %}
+      path = name_to_path name
+      read_proc_sys(path).to_i32
+    {% else %}
+      raise NotImplemented("Sysctl", "is not supported on this platform")
```
Comment (Sija) [inline src/sysctl.cr]: Missing `.new`.

File: src/exception.cr [hunk 4]
```diff
@@ -91,3 +91,16 @@ class DivisionByZero < Exception
...
+# Raised when a feature is not implemented yet or when it cannot be implemented
+# on a certain system or platform
+#
+# ```
+# Sysctl.get_int("net.ipv4.forward") # Sysctl doesn't exist on Windows
+# ````
+#
+class NotImplemented < Exception
+  def initialize(what : String, reason = "is not implemented")
+    super [what, reason].join(" ")
```
Comment (Sija) [inline src/exception.cr]: Use `Tuple` instead of an `Array`: `{what, reason}.join(' ')`

File: src/sysctl.cr [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    def name_to_path(name : String)
+      "/proc/sys/" + name.gsub(".", "/")
+    end
+
+    # :nodoc:
+    def read_proc_sys(path : String) : String
+      if File.readable?(path)
+        File.read(path)
+      else
+        raise Errno.new("Unable to read #{path}", Errno::ENOENT)
```
Comment (Sija) [inline src/sysctl.cr]: Shouldn't this be just `Exception`?

File: src/sysctl.cr [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      value_len = GC.malloc_atomic(sizeof(Int32)).as(Int32*)
+      value_len.value = max_size
+
+      res = LibC.sysctlbyname(name.to_unsafe, value.as(Void*),
+        value_len.as(LibC::SizeT*), Pointer(Void).null, 0)
+
+      if res == 0
+        String.new(value, value_len.value - 1, 1)
+      else
+        raise Errno.new("Unable to fetch sysctl value #{name}")
```
Comment (Sija) [inline src/sysctl.cr]: Shouldn't this be just `Exception`?

File: src/sysctl.cr [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  # Returns the sysctl named *name* as an integer. Will raise `NotImplemented`
+  # if the sysctl interface doesn't exist on your system (Windows) or isn't
+  # supported yet. Will raise the appropriate `Errno` exception in case of
+  # error.
+  #
+  # ```
+  # Sysctl.get_int("hw.ncpu") # => 8
+  # ```
+  #
+  def get_i32(name : String) : Int32
```
Comment (Sija) [inline src/sysctl.cr]: I'd drop `get_` prefix for consistency with Ruby/Crystal API expectations (namely no accessor - `get_`, `set_`, `is_` - prefixes).

File: src/sysctl.cr [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+  # Returns a sysctl value named *name* as a `String`. Will raise
+  # `NotImplemented` on systems where the sysctl interface doesn't exists or
+  # isn't supported. Will also raise the appropriate `Errno` in case of error.
+  #
+  # ```
+  # Sysctl.get_str("kernel.random.boot_id", 42) # => "1c629298-a60b-48df-bba1-9ea77f6b37ba"
+  # ```
+  #
+  def get_str(name : String, max_size : Int32) : String
```
Comment (Sija) [inline src/sysctl.cr]: ditto

File: src/exception.cr [hunk 9]
```diff
@@ -91,3 +91,16 @@ class DivisionByZero < Exception
     super(message)
   end
 end
+
+# Raised when a feature is not implemented yet or when it cannot be implemented
+# on a certain system or platform
+#
+# ```
+# Sysctl.get_int("net.ipv4.forward") # Sysctl doesn't exist on Windows
```
Comment (RX14) [inline src/exception.cr]: Typically we use `# raises ExceptionType (message)` when documenting raising in examples [(ref)](https://github.com/crystal-lang/crystal/blob/master/src/io/memory.cr#L380)

File: src/sysctl.cr (hunk 1 above)
Comment (RX14) [inline src/sysctl.cr]: See: https://github.com/crystal-lang/crystal/pull/4237

File: src/sysctl.cr (hunk 7 above)
Comment (RX14) [inline src/sysctl.cr]: What would you name setters? I don't see a good alternative here.

File: src/sysctl.cr (hunk 1 above)
Comment (elthariel) [inline src/sysctl.cr]: I thought about it, but it would be to a lot of duplication, as the implementation would be the same for osx/freebsd and all the linux variants.
Also it was implemented this way in src/process/executable_path.cr

Should I still move it there ?

File: src/sysctl.cr (hunk 7 above)
Comment (elthariel) [inline src/sysctl.cr]: makes perfect sense

File: src/sysctl.cr (hunk 3 above)
Comment (elthariel) [inline src/sysctl.cr]: -_- Thanks

File: src/sysctl.cr (hunk 7 above)
Comment (Sija) [inline src/sysctl.cr]: @RX14 
```cr
# getter
def i32(name : String) : Int32; end
# setter
def i32(name : String, value : Int32) : Void; end
```

File: src/sysctl.cr (hunk 1 above)
Comment (RX14) [inline src/sysctl.cr]: Sorry for the misinformation! as this is in `lib_c` you don't need the `skip` macro.

File: src/sysctl.cr (hunk 6 above)
Comment (elthariel) [inline src/sysctl.cr]: `sysctl` sets the `errno` in case of error, so I think we should propagate it.

File: src/sysctl.cr (hunk 5 above)
Comment (elthariel) [inline src/sysctl.cr]: We might use the standard exception here but it would be inconsistent with the other system's implementation where we probably want to libc `errno` to be reported to the user.

File: src/exception.cr (hunk 4 above)
Comment (elthariel) [inline src/exception.cr]: Roger

File: src/exception.cr (hunk 9 above)
Comment (elthariel) [inline src/exception.cr]: Thanks

File: src/sysctl.cr (hunk 1 above)
Comment (elthariel) [inline src/sysctl.cr]: Shall I move this in the arch-dependant folders ?

File: src/sysctl.cr (hunk 7 above)
Comment (elthariel) [inline src/sysctl.cr]: *Perfetto!*

**Comment by elthariel** [comment]
Thanks to both of you for the quick and insightful review

File: spec/std/sysctl_spec.cr [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+describe Sysctl do
+  describe "get_int" do
+    it "returns the sysctl value as an int" do
+      keys = {% if flag?(:darwin) %}
+        ["hw.ncpu", "hw.byteorder"]
+      {% elsif flag?(:linux) %}
+        ["net.ipv4.ip_forward", "kernel.pty.max"]
+      {% else %}
+        []
```
Comment (Sija) [inline spec/std/sysctl_spec.cr]: Should be `[] of String` (see [failed travis build](https://travis-ci.org/crystal-lang/crystal/jobs/227197195#L168)).

File: src/sysctl.cr (hunk 1 above)
Comment (RX14) [inline src/sysctl.cr]: Yes. The bindings in those folders should actually be automatically generated from the `.h` files on each platform, but I never got [the generator](https://github.com/ysbaddaden/posix) to work. @ysbaddaden?

File: src/sysctl.cr (hunk 7 above)
Comment (RX14) [inline src/sysctl.cr]: @Sija Is there any precedence for that naming in the stdlib? I think it's terrible.

Comment (Sija) [inline src/sysctl.cr]: I've seen API like that in [kemal-session](https://github.com/kemalcr/kemal-session) shard for instance.

File: src/sysctl.cr (hunk 6 above)
Comment (RX14) [inline src/sysctl.cr]: (I think `Errno` is terrible design since it exposes platform-specifics at a high level, and on a more practical level you can't `rescue` only a single errno without using `if` inside `rescue`, however i'm too lazy to make an RFC)

Comment (oprypin) [inline src/sysctl.cr]: Just copy [this](https://docs.python.org/3/library/exceptions.html#os-exceptions) and you've got an RFC 😜

File: src/sysctl.cr (hunk 1 above)
Comment (elthariel) [inline src/sysctl.cr]: That's fun. I hacked something similar for OpenGL

Comment (elthariel) [inline src/sysctl.cr]: Anyway, the variants of sysctl (`sysctl_by_name`) I'm using isn't really POSIX compliant. Plus this call is marked as deprecated on Linux, so how would that play with this generator ?

File: src/sysctl.cr (hunk 6 above)
Comment (elthariel) [inline src/sysctl.cr]: IMHO, one of the point of Crystal is to allow system programming as well, which requires exposing platform specific details. That being said, I understand this should be allowed, not forced.

File: src/sysctl.cr (hunk 7 above)
Comment (elthariel) [inline src/sysctl.cr]: I've seen and done that in Ruby (and python fwiw) as well

File: src/sysctl.cr (hunk 1 above)
Comment (ysbaddaden) [inline src/sysctl.cr]: I second that anything LibC should be in the arch specific under `src/lib_c` with file namings following the actual C headers; thus way we may be capable in the future to generate these on-the-flt.

We prefer POSIX, but don't enforce it, if it requires some specific but public API for a platform, or the function is so common across platforms that it can be relied on, you can use it.

That being said, if something is deprecated please don't use it; it will likely break some day.

File: src/exception.cr [hunk 11]
```diff
@@ -91,3 +91,16 @@ class DivisionByZero < Exception
...
+#
+# ```
+# Sysctl.get_int("net.ipv4.forward") # raises NotImplemented ("Sysctl is not implemented on that platform")
+# ````
+#
+class NotImplemented < Exception
+  def initialize(what : String, reason = "is not implemented")
+    super {what, reason}.join(" ")
+  end
+end
```
Comment (ysbaddaden) [inline src/exception.cr]: No, this should fail to compile, not fail at runtime. One solution is to not expose the Sysctl type on unsupported platforms.

File: src/sysctl.cr (hunk 1 above)
Comment (elthariel) [inline src/sysctl.cr]: @ysbaddaden, The syscall is deprecated on Linux but it's not available through the libc anymore, so no risks there. So your suggestion is to write those by hands until it is supported by the generator, am I right ?

File: src/exception.cr (hunk 11 above)
Comment (elthariel) [inline src/exception.cr]: Sure

**Comment by elthariel** [comment]
@ysbaddaden, Even if it Linux deprecated it, `sysctl`is still part of POSIX and libc and generally widely available on Unix. 
Also, I'm not sure I understand Crystal goal here. By providing things like pointers and std access to libc, it kind of positions itself to some extent as a *system language*. And if you're a system language, it seems to me that you want access to sysctl and ioctl, no ? It also doesn't prevent us from building nicer abstractions. 

That being said, I might be mistaken about it, and maybe it's better to do occasional on-the-spot wrappings like here: https://github.com/crystal-lang/crystal/blob/master/src/process/executable_path.cr#L74

It's up to you and the community. What would be the process to reach a verdict here ?

File: src/exception.cr (hunk 11 above)
Comment (elthariel) [inline src/exception.cr]: Shall we keep the NotImplemented exception ? (like ruby & python for example), should I move it to a separate PR ? or should I get rid of it ?

Comment (ysbaddaden) [inline src/exception.cr]: No. It may be meaningful in interpreted languages, not in a compiled language: not implemented => compilation failure.

Comment (elthariel) [inline src/exception.cr]: Duh. Now you're pointing at it...
Would only be useful with dlopen & friends. I'll get rid of it tomorrow

**Comment by ysbaddaden** [comment]
Pointers are required to bind to C libraries, they speed up implementations, both internally (for core/stdlib) and externally (for shards, apps); and without them we wouldn't achieve much, if anything. On the other hand, `sysctl` and `ioctl` are platform specific (it's *not* POSIX) and not *required* to build anything on, but specific use cases. We want a wide stdlib, covering general use cases, just not that wide and specific :)

**Comment by elthariel** [comment]
Hum... This is embarrassing.  Just confirmed with the Single Unix Specification v4, `sysctl` isn't mentionned anywhere in the standard. I always thought it was part of SUS. As a sidenote,`ioctl` is defined by SUS/POSIC (but it's not the topic)

@ysbaddaden, To get back on the initial topic, I think `sysctl` is still the easier way to get the number of cores/cpus on OSX. So you'd advocate in favor of a *hidden* binding when it's needed and provide an higher level API for that ?

**Comment by ysbaddaden** [comment]
Do you have more examples and use cases that would benefit everyone? I mean other than the number of available CPU?

**Comment by elthariel** [comment]
I haven't really thought about it. That's an API I use from times on times with C and since I needed it for the number of cpu on OSX, I thought it would be a useful contribution. I'll have a look and think a bit about it if it can gives input for the community to decide whether it is or not

**Comment by elthariel** [comment]
I should precise that I'm more familiar with the Linux system APIs, which aren't based out of sysctl anymore, outside of a few idiomatic things you'll find here and here.

On the OSX side of things, it seems like a good way to get access to a bunch of system data about shared memory, VM, networking and hardware. I'm no OSX experts, so there might be better way to get access to this. We might also need input from OpenBSD and FreeBSD guys, but I think that might be the only way to access a whole bunch of similar system data. I think FreeBSD started to move to something similar to procfs, but I don't think OpenBSD was on that path (although I must confess the last time I worked with awesome openbsd was quite a while ago).

I certainly don't want to push on this topic, as you made pretty good points as of how this wouldn't be the most widely useful API. On the other hand, we'll need at least some portion of it internally so why not expose it ? TBH, I don't have anywhere near enough vision on Crystal's community and it's goals to have a strong opinion. You'll have to place the threshold :)

**Comment by mverzilli** [comment]
As @ysbaddaden said, this seems a very good fit for a separate shard:

1. It wraps a system specific tool.
2. Correct me if I'm wrong, but it doesn't seem to need "first class" compiler support or modifications.
3. We aren't sure if we want this in the stdlib yet, and would probably like to approach it based on use cases in a more platform agnostic way.

So, I'd put it in a shard, and if the ecosystem embraces it and it becomes a very common item in `shard.yml` files there's always time for the stdlib to "absorb" it.

**Comment by elthariel** [comment]
@mverzilli So be it. I'm not sure as I'm going to effectively create the shard as having it out of stdlib defeats the initial purpose of it.
Regarding this initial use case, I'm not sure there's a platform agnostic way to do it (i think it's quite the contrary), so we'd still have to implement something like this hidden somewhere
